keyutils/keyutils.go

```diff
@@ -0,0 +1,87 @@
+package keyutils
+
+import (
+	"crypto"
+	cryptorand "crypto/rand"
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"os"
+
+	"github.com/cloudflare/cfssl/helpers"
+	"github.com/docker/swarmkit/pkcs8"
+)
+
+// FIPSEnvVar is the environment variable name for FIPS mode
+const FIPSEnvVar = "GOFIPS"
+
+// FIPSEnabled returns true if FIPS mode is enabled
+func FIPSEnabled() bool {
+	return os.Getenv(FIPSEnvVar) != ""
+}
+
+func isPKCS8(pemBytes []byte) bool {
+	if _, err := x509.ParsePKCS8PrivateKey(pemBytes); err == nil {
+		return true
+	}
+
+	return pkcs8.IsEncryptedPEMBlock(&pem.Block{
+		Type:    "PRIVATE KEY",
+		Headers: nil,
+		Bytes:   pemBytes,
+	})
+}
+
+// ParsePrivateKeyPEMWithPassword parses PKCS#1 (except in FIPS mode) or PKCS#8 PEM with password to crypto.Signer
+func ParsePrivateKeyPEMWithPassword(pemBytes, password []byte) (crypto.Signer, error) {
+	block, _ := pem.Decode(pemBytes)
+	if block == nil {
+		return nil, errors.New("Could not parse PEM")
+	}
+
+	if isPKCS8(block.Bytes) {
+		return pkcs8.ParsePrivateKeyPEMWithPassword(pemBytes, password)
+	} else if FIPSEnabled() {
+		return nil, errors.New("Unsupported key format")
+	}
+
+	return helpers.ParsePrivateKeyPEMWithPassword(pemBytes, password)
+}
+
+// IsEncryptedPEMBlock checks if a PKCS#1 (except in FIPS mode) or PKCS#8 PEM-block is encrypted or not
+func IsEncryptedPEMBlock(block *pem.Block) bool {
+	if isPKCS8(block.Bytes) {
+		return pkcs8.IsEncryptedPEMBlock(block)
+	} else if FIPSEnabled() {
+		panic("Unsupported key format")
```

Comment:
I don't think this panic is the right thing to do if we encounter an unsupported key format. The function should probably return an error in this case.